Add methods for retrieving formats applied to a range.

willrowe opened the pull request:
This PR adds three methods: `getFormats`, `getLeafFormats`, and `getLineFormats` as discussed in #379. They all accept either a start and end index, a range, or if nothing is passed then the current selection range is used. It will then return an object representing all the formats and values which have been applied to the **entire** range. These methods are based on the ones in the [toolbar module](https://github.com/quilljs/quill/blob/54cae74b1994d87bea7fb349733abdc9ddb0c697/src/modules/toolbar.coffee#L105-L130), but differ in that they only return those formats which extend over the whole range. For example, if two different fonts are set for a range or only part of the range is set to bold then neither of those formats will be returned.
This is a work in progress, so I'm looking for feedback to improve the functionality. For instance, I'm a little unsure about the name of the methods, especially `getFormats` as that might be confusing.

willrowe commented:
As far as naming goes, `getLeafFormats` and `getLineFormats` are pretty explicit in that they imply that they are formats for a certain part of the content (whether it be a leaf or a line). `getFormats` could be confusing because it would seem that it should return a list of all the formats that are currently registered with Quill. Maybe `getTextFormats` would be more clear?
Another alternative would be to stick with how the toolbar module was doing it and call them `activeFormats`, `activeLeafFormats`, and `activeLineFormats`.

thomsbg commented:
I like `getActiveFormats`. I don't fully understand the purpose of exposing `activeLeafFormats` and `activeLineFormats` separately as public methods.

willrowe commented:
The main method returns both leaf and line formats, but if only one or the other is required then you'd want access to those methods. These are mostly to expose methods for use with external modules and creating a more robust API. Does that make sense?

jhchen commented:
I'd worry about adding 3 new API calls to the namespace as well. They could be merged into one and just take an optional parameter to exclude line/leaf formats.

willrowe commented:
I'll make those other two private then. A third parameter would determine what would be returned and default to getting both, should I add constants for these values? Like `ALL`, `LEAF`, and `LINE`?

HHogg commented on src/quill.coffee:
```diff
@@ -241,6 +282,13 @@ class Quill extends EventEmitter2
     return params
 
+  _intersectFormats: (formatGroups) ->
+    return _.reduce(formatGroups.slice(1), (inAll, formats) ->
```
Hi @willrowe, I was in need of this today, so thank you. 

Have tried it out but I have an issue when calling the get methods when the selection spans across formatted and unformatted text, which leaves `formats` `undefined`. Just need a `formats = {}` here? Similar to the `Toolbar` module implementation. 

```coffee
return _.reduce(formatGroups.slice(1), (inAll, formats = {}) ->
```

willrowe commented:
Thanks for pointing that out @HHogg. I've added tests for that situation and fixed it. I also decided to not add the option to specify what kinds of formats should be returned right now. This simplifies it to bare bones.

willrowe commented:
Any more feedback for this?

jhchen commented:
It'd be much more useful if these calls returned the active formats--that is if I type a letter, these formats will be applied to that letter. They may be visible or invisible to the DOM which makes it tricky to implement. For example if I hit Cmd+B to activate bold, nothing changes in the DOM. Yet calling these methods at this point should give me this information. The toolbar does implement this but bugs crop up all the time on edge cases on different browsers so the ideal implementation would for one not require the toolbar module and two be very reliable on all supported platforms.

willrowe commented:
That would be very nice, seeing as now it just grabs it from the preceding character. That would require more work in other areas, however, so I think it'd be outside the scope of this particular PR. If that's developed it could definitely be tied into this function later on.

pedrosanta commented on src/quill.coffee:
```diff
@@ -241,6 +253,36 @@ class Quill extends EventEmitter2
+    return this._intersectFormats(formats)
+
+  _intersectFormats: (formatGroups) ->
+    return _.reduce(formatGroups.slice(1), (inAll, formats) ->
```
Could you explain it better why you are doing `formatGroups.slice(1)` here?

It seems to be ignoring the active formats of the first 'leaf', right? Shouldn't those first leaf formats be included/returned as well?

pedrosanta replied:
Ok, just took notice that this is supposed to return formats applied to the **whole** range.

pedrosanta replied:
Probably because it was copied out from https://github.com/quilljs/quill/blob/develop/src/modules/toolbar.coffee#L133...

willrowe replied:
@pedrosanta if you look at line 284, the first leaf is passed in there. This is because the `reduce` function requires a starting point, so the first leaf it broken off and used as the base and then all subsequent leaves are compared on top of it.

pedrosanta replied:
@willrowe Oh, ok, makes sense. Thanks. :grinning:

pedrosanta commented:
@willrowe yhea, regarding the name, I also feel it might be somewhat confusing. For instance, I assumed this methods were to obtain all the active formats within a range (either if they matched the whole, or only a part of the range). Don't know if there's some sort of functionality around for that.

willrowe commented:
@pedrosanta that's something that I'm not sure about. It would depend on the use case. I can see the following situations:
- Getting the formats that are applied to an entire range.
- Getting any formats in a range.
- Determining whether a range has a certain format applied to it.
Should there be multiple methods to cover these situations or just changing the name back to `getFormats` and having arguments to change what is returned?

thomsbg commented:
Remember also that quill.getContents(start, end) allows you to inspect the rich text delta data structure for a range, which includes details of which formats are applied at the most granular level. You could write your own function to find the intersection or union of formats that way.

jhchen commented:
Since this PR was opened to solicit feedback and feedback has been given. If there are further ambitions to take on the suggestions feel free to open another PR.

willrowe commented:
I'm planning on rewriting it using some of the notes I gathered through this PR. I'll either re-submit after parchment has been merged or 1.0 has been released.
